File: apps/python-api/app/store/timescale.py
from datetime import datetime, timedelta, timezone

import asyncpg

from app.models import Sample, Tag

_LOCF_LOOKBACK = timedelta(hours=3)

_LOCF_SQL = """
SELECT s.ts, s.tag_id, s.value, s.quality
FROM unnest($1::int4[]) AS t(tag_id)
CROSS JOIN LATERAL (
    SELECT ts, tag_id, value, quality
    FROM samples
    WHERE samples.tag_id = t.tag_id AND ts <= $2 AND ts >= $3
    ORDER BY ts DESC
    LIMIT 1
) s
"""

_LOCF_UNBOUNDED_SQL = """
SELECT s.ts, s.tag_id, s.value, s.quality
FROM unnest($1::int4[]) AS t(tag_id)
CROSS JOIN LATERAL (
    SELECT ts, tag_id, value, quality
    FROM samples
    WHERE samples.tag_id = t.tag_id AND ts <= $2
    ORDER BY ts DESC
    LIMIT 1
) s
"""

_RANGE_SQL = """
SELECT ts, tag_id, value, quality
FROM samples
WHERE tag_id = ANY($1) AND ts > $2 AND ts <= $3
ORDER BY tag_id, ts
"""


def _utc(ts: datetime) -> datetime:
    if ts.tzinfo is None:
        return ts.replace(tzinfo=timezone.utc)
    return ts.astimezone(timezone.utc)


def _merge_range(tag_ids: list[int], head: list[Sample], tail: list[Sample]) -> list[Sample]:
    buckets: dict[int, list[Sample]] = {int(t): [] for t in tag_ids}
    extra: list[Sample] = []
    for sample in head:
        buckets.setdefault(sample.tag_id, []).append(sample)
    for sample in tail:
        if sample.tag_id in buckets:
            buckets[sample.tag_id].append(sample)
        else:
            extra.append(sample)
    out: list[Sample] = []
    for tag_id in tag_ids:
        out.extend(buckets[int(tag_id)])
    out.extend(extra)
    return out


class TimescaleStore:
    name = "timescaledb"

    def __init__(self, dsn: str) -> None:
        self._dsn = dsn
        self._pool: asyncpg.Pool | None = None

    async def _conn(self) -> asyncpg.Pool:
        if self._pool is None:
            self._pool = await asyncpg.create_pool(
                self._dsn,
                min_size=4,
                max_size=16,
                server_settings={"TimeZone": "UTC"},
            )
        return self._pool

    async def ping(self) -> None:
        pool = await self._conn()
        await pool.execute("SELECT 1")

    async def write(self, samples: list[Sample]) -> None:
        if not samples:
            return
        pool = await self._conn()
        records = [
            (_utc(s.ts), int(s.tag_id), float(s.value), int(s.quality)) for s in samples
        ]
        await pool.copy_records_to_table(
            "samples",
            records=records,
            columns=["ts", "tag_id", "value", "quality"],
        )

    async def locf(self, tag_ids: list[int], at: datetime) -> list[Sample]:
        at = _utc(at)
        ids = [int(i) for i in tag_ids]
        rows = await self._locf(ids, at, True)
        found = {s.tag_id for s in rows}
        missing = [i for i in ids if i not in found]
        if missing:
            rows.extend(await self._locf(missing, at, False))
        return rows

    async def _locf(self, tag_ids: list[int], at: datetime, bounded: bool) -> list[Sample]:
        pool = await self._conn()
        if bounded:
            raw = await pool.fetch(_LOCF_SQL, tag_ids, at, at - _LOCF_LOOKBACK)
        else:
            raw = await pool.fetch(_LOCF_UNBOUNDED_SQL, tag_ids, at)
        return [
            Sample(
                ts=_utc(r["ts"]), tag_id=int(r["tag_id"]), value=float(r["value"]), quality=int(r["quality"])
            )
            for r in raw
        ]

    async def range(self, tag_ids: list[int], start: datetime, end: datetime) -> list[Sample]:
        start, end = _utc(start), _utc(end)
        ids = [int(i) for i in tag_ids]
        head = [s.as_carried() for s in await self.locf(ids, start)]
        pool = await self._conn()
        raw = await pool.fetch(_RANGE_SQL, ids, start, end)
        tail = [
            Sample(
                ts=_utc(r["ts"]), tag_id=int(r["tag_id"]), value=float(r["value"]), quality=int(r["quality"])
            )
            for r in raw
        ]
        return _merge_range(ids, head, tail)

    async def upsert_tags(self, tags: list[Tag]) -> None:
        pool = await self._conn()
        await pool.executemany(
            """
            INSERT INTO tags (id, name, unit) VALUES ($1, $2, $3)
            ON CONFLICT (id) DO UPDATE SET name = EXCLUDED.name, unit = EXCLUDED.unit
            """,
            [(t.id, t.name, t.unit) for t in tags],
        )

    async def list_tags(self) -> list[Tag]:
        pool = await self._conn()
        rows = await pool.fetch("SELECT id, name, COALESCE(unit, '') AS unit FROM tags ORDER BY id")
        return [Tag(id=r["id"], name=r["name"], unit=r["unit"]) for r in rows]

    async def close(self) -> None:
        if self._pool is not None:
            await self._pool.close()
            self._pool = None

File: apps/go-api/internal/store/timescale.go
package store

import (
	"context"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"prism/go-api/internal/model"
)

type Timescale struct {
	pool *pgxpool.Pool
}

func NewTimescale(dsn string) (*Timescale, error) {
	cfg, err := pgxpool.ParseConfig(dsn)
	if err != nil {
		return nil, err
	}
	cfg.MaxConns = 16
	cfg.MinConns = 4
	pool, err := pgxpool.NewWithConfig(context.Background(), cfg)
	if err != nil {
		return nil, err
	}
	return &Timescale{pool: pool}, nil
}

func (s *Timescale) Name() string { return "timescaledb" }

func (s *Timescale) Close() error {
	s.pool.Close()
	return nil
}

func (s *Timescale) Ping(ctx context.Context) error {
	return s.pool.Ping(ctx)
}

// Archive max gap is 1h; 3h still finds the previous minute/hour point at 364d ago.
const tsLocfLookback = 3 * time.Hour

const tsLocfSQL = `
	SELECT s.ts, s.tag_id, s.value, s.quality
	FROM unnest($1::int4[]) AS t(tag_id)
	CROSS JOIN LATERAL (
		SELECT ts, tag_id, value, quality
		FROM samples
		WHERE samples.tag_id = t.tag_id AND ts <= $2 AND ts >= $3
		ORDER BY ts DESC
		LIMIT 1
	) s`

const tsLocfUnboundedSQL = `
	SELECT s.ts, s.tag_id, s.value, s.quality
	FROM unnest($1::int4[]) AS t(tag_id)
	CROSS JOIN LATERAL (
		SELECT ts, tag_id, value, quality
		FROM samples
		WHERE samples.tag_id = t.tag_id AND ts <= $2
		ORDER BY ts DESC
		LIMIT 1
	) s`

const tsRangeSQL = `
	SELECT ts, tag_id, value, quality
	FROM samples
	WHERE tag_id = ANY($1) AND ts > $2 AND ts <= $3
	ORDER BY tag_id, ts`

func (s *Timescale) Write(ctx context.Context, samples []model.Sample) error {
	if len(samples) == 0 {
		return nil
	}
	_, err := s.pool.CopyFrom(
		ctx,
		pgx.Identifier{"samples"},
		[]string{"ts", "tag_id", "value", "quality"},
		&sampleCopySrc{samples: samples, idx: -1, row: make([]any, 4)},
	)
	return err
}

type sampleCopySrc struct {
	samples []model.Sample
	idx     int
	row     []any
}

func (s *sampleCopySrc) Next() bool {
	s.idx++
	return s.idx < len(s.samples)
}

func (s *sampleCopySrc) Values() ([]any, error) {
	p := &s.samples[s.idx]
	s.row[0] = p.TS.UTC()
	s.row[1] = int32(p.TagID)
	s.row[2] = float32(p.Value)
	s.row[3] = int16(p.Quality)
	return s.row, nil
}

func (s *sampleCopySrc) Err() error { return nil }

func (s *Timescale) Locf(ctx context.Context, tagIDs []uint32, at time.Time) ([]model.Sample, error) {
	out, err := s.locf(ctx, tagIDs, at, true)
	if err != nil {
		return nil, err
	}
	missing := missingTagIDs(tagIDs, out)
	if len(missing) == 0 {
		return out, nil
	}
	rest, err := s.locf(ctx, missing, at, false)
	if err != nil {
		return nil, err
	}
	return append(out, rest...), nil
}

func (s *Timescale) locf(ctx context.Context, tagIDs []uint32, at time.Time, bounded bool) ([]model.Sample, error) {
	at = at.UTC()
	var (
		rows pgx.Rows
		err  error
	)
	if bounded {
		rows, err = s.pool.Query(ctx, tsLocfSQL, intTags(tagIDs), at, at.Add(-tsLocfLookback))
	} else {
		rows, err = s.pool.Query(ctx, tsLocfUnboundedSQL, intTags(tagIDs), at)
	}
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return scanSamples(rows, false)
}

func (s *Timescale) Range(ctx context.Context, tagIDs []uint32, from, to time.Time) ([]model.Sample, error) {
	head, err := s.Locf(ctx, tagIDs, from)
	if err != nil {
		return nil, err
	}
	rows, err := s.pool.Query(ctx, tsRangeSQL, intTags(tagIDs), from.UTC(), to.UTC())
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	tail, err := scanSamples(rows, false)
	if err != nil {
		return nil, err
	}
	return mergeCHRange(tagIDs, head, tail), nil
}

func (s *Timescale) UpsertTags(ctx context.Context, tags []model.Tag) error {
	batch := &pgx.Batch{}
	for _, t := range tags {
		batch.Queue(`
			INSERT INTO tags (id, name, unit) VALUES ($1, $2, $3)
			ON CONFLICT (id) DO UPDATE SET name = EXCLUDED.name, unit = EXCLUDED.unit`,
			int32(t.ID), t.Name, t.Unit)
	}
	br := s.pool.SendBatch(ctx, batch)
	defer br.Close()
	for range tags {
		if _, err := br.Exec(); err != nil {
			return err
		}
	}
	return nil
}

func (s *Timescale) ListTags(ctx context.Context) ([]model.Tag, error) {
	rows, err := s.pool.Query(ctx, `SELECT id, name, COALESCE(unit, '') FROM tags ORDER BY id`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []model.Tag
	for rows.Next() {
		var t model.Tag
		var id int32
		if err := rows.Scan(&id, &t.Name, &t.Unit); err != nil {
			return nil, err
		}
		t.ID = uint32(id)
		out = append(out, t)
	}
	return out, rows.Err()
}

func intTags(ids []uint32) []int32 {
	out := make([]int32, len(ids))
	for i, id := range ids {
		out[i] = int32(id)
	}
	return out
}

func scanSamples(rows pgx.Rows, carried bool) ([]model.Sample, error) {
	out := make([]model.Sample, 0, 16)
	for rows.Next() {
		var s model.Sample
		var tag int32
		var q int16
		if err := rows.Scan(&s.TS, &tag, &s.Value, &q); err != nil {
			return nil, err
		}
		s.TagID = uint32(tag)
		s.Quality = uint16(q)
		s.Carried = carried
		out = append(out, s)
	}
	return out, rows.Err()
}

